Address rootless-x11 cases

aleixpol opened the pull request:
The notable changes are:
* Stop the greeter before we start the new session
* Properly accounting for Xorg's -novtswitch
* Don't start a compositor after a session is closed, we were leaking them (x11-user & wayland)
* Improve debugging

Fixes https://github.com/sddm/sddm/issues/1456

aleixpol committed:
Remove unnecessary headers

aleixpol committed:
x11user: Need to switch to its vt as it starts

We are launching Xorg with -novtswitch, so we need to explicitly jump to
its vt, otherwise it doesn't happen.

aleixpol committed:
Display: Do not start the greeter from the display ends

The stop() call will trigger a stop and start of the Display entirely
(see Seat::displayStopped). Therefore here we are only starting a
greeter that was never fully set up.

aleixpol committed:
Display: stop the greeter before we start the session

This way we make sure they don't overlap each other.

aleixpol committed:
UserSession: Improve "Failed to take control" error message

aleixpol committed:
Use a compile-time connect for SocketServer

aleixpol committed:
MessageHandler: Ensure debug files are written

If the user can't write in /var/log/sddm.log, it will just be ignored
otherwise.

aleixpol committed:
Display: Use a QPointer for the local socket

It's not owned by us and it can be destroyed at times.

aleixpol committed:
helper: Make sure helper processes terminate upon SIGTERM

aleixpol committed:
UserSession: Don't stop the process if it was never started

evelikov commented:
Coming from https://github.com/sddm/sddm/issues/1471#issuecomment-968331648 - two small requests:
 - considering this is considered a blocker, let's swap the "polish" with something more accurate
 - can you elaborate what kind rounds should we do - is there anything specific to look for?
 - non-rootful -> rootless is the more common term

evelikov commented on src/common/MessageHandler.h:
```diff
@@ -75,6 +77,14 @@ namespace SDDM {
                 file.open(QFile::Truncate | QFile::WriteOnly);
         }
 
+        // If we can't open the file, create it in a writable location
```
For the non-QT experts, can you provide an inline example where this "writable location" might be?

aleixpol replied:
You can test it yourself:
$ qtpaths --writable-path AppDataLocation
/home/apol/.local/share/<APPNAME>

evelikov replied:
Neat, can you add an inline example?

aleixpol replied:
Added examples on the comment.

evelikov commented on src/helper/HelperSignalHandler.h:
```diff
@@ -0,0 +1,36 @@
+// ensure we exit gracefully and close the session if sigtermed (i.e restarting sddm)
+static void sigtermHandler(int signalNumber)
+{
+    Q_UNUSED(signalNumber)
```
~~Both commit message and comment says "SIGTERM" yet we ignore the signal all together - SIGTERM or otherwise.~~
~~Why is that - can we tweak the comment to mention the reason?~~

Edit: brain fart - we get here only on SIGTERM, as seen in the `signal(SIGTERM, sigtermHandler)` line below

evelikov commented on src/helper/UserSession.cpp:
```diff
@@ -100,17 +100,18 @@ namespace SDDM {
 
     void UserSession::stop()
```
Silly question (last one I promise) - what's the problem with calling `terminate()`?

It seems that the more notable change is that we omit `waitForFinished()` will be blocking for potentially notable amount of time.

davidedmundson replied:
we do still call terminate and waitForFinished() we've just wrapped things in an if statement.

evelikov replied:
Right, I meant - what's the problem with calling `terminate()` and friends like `waitForFinished()`) in the `state() == QProcess::NotRunning` case?

aleixpol replied:
We first call terminate which is more gentle, if terminate doesn't make the project finish, we kill it.

davidedmundson commented:
>can you elaborate what kind rounds should we do - is there anything specific to look for?

Every combo of:
autologin, root X11 greeter, rootless X11 greeter, wayland greeter
going into 
X11 session, Wayland session

including log out and going between them.

If you get anything explode or a black screen, it's not working great.

davidedmundson commented on src/daemon/Display.cpp:
```diff
@@ -466,10 +467,6 @@ namespace SDDM {
         if (status != Auth::HELPER_AUTH_ERROR)
             stop();
-
-        // Start the greeter again as soon as the user session is closed
```
so what happens now when a user does log out in a root-full X session?

aleixpol replied:
Display::stop gets called (right before the removal). This informs Seat that we've stooped and we go to Seat::displayStopped.

This removes the display and if it's display-less (which is the case as we just removed the display) it creates a new one which will start a new greeting process.

aleixpol committed:
Make sure we handle UNIX signals properly

It's very unsafe to interact with the event loop from a signal callback,
so we use internal sockets to handle it.
For more information see:
https://doc.qt.io/qt-5/unix-signals.html

davidedmundson commented on src/common/CustomSignalHandler.cpp:
```diff
@@ -0,0 +1,60 @@
+
+int CustomSignalHandler::signalFd[2];
+
+CustomSignalHandler::CustomSignalHandler()
```
We also have src/daemon/SignalHandler.cpp:

which is near identical to this class. Only more verbose


We don't want both.
 
Can we either port:
src/daemon/DaemonApp.cpp:        connect(m_signalHandler, &SignalHandler::sigintReceived, this, &DaemonApp::quit);
src/daemon/DaemonApp.cpp:        connect(m_signalHandler, &SignalHandler::sigtermReceived, this, &DaemonApp::quit);

to this and drop SignalHandler.h or vice-versa.

aleixpol replied:
Fair enough, added custom signal handling on the previous one so that we changed the least and moved it to common.

This class could use a cleanup, there's quite a bit of unneeded duplicate code as far as I understand, probably not for this MR though.

aleixpol committed:
Display: Do not set up the VT twice

Also remove unnecessary copy of the Session.
Make sure m_terminalId is always set up to the correct value.
